### PR TITLE
Enrike Valentinov Homework 2

### PR DESCRIPTION
**Pull request by Enrike04**

### CODE REVIEW

File: Main.hs [hunk 1]
```diff
@@ -11,28 +11,47 @@ import qualified Data.List as L
...
 multByIndex :: [Int] -> [Int]
-multByIndex = undefined
+multByIndex xs = zipWith (*) xs [0..]
 
 powerByIndex :: [Int] -> [Int]
-powerByIndex = undefined
+powerByIndex xs = zipWith (^) xs [0..]
 
+prodList [] = 1
+prodList (h:t) = h * prodList t
```
Comment (alex28sh) [inline Main.hs:20]: you should use functions discussed in lecture instead this

File: Main.hs [hunk 2]
```diff
@@ -11,28 +11,47 @@ import qualified Data.List as L
...
+
+short :: [a] -> Bool
+short [] = True
+short [_] = True
+short _ = False
```
Comment (alex28sh) [inline Main.hs:25]: this check can be done in 2 lines (also you don't need it in separate function)

File: Main.hs [hunk 3]
```diff
@@ -11,28 +11,47 @@ import qualified Data.List as L
...
+isSorted xs = and (zipWith (<=) xs (tail xs))
 
-countElement :: Int -> [Int] -> Int 
-countElement = undefined
+countElement :: Int -> [Int] -> Int
+countElement x [] = 0
+countElement x (y:ys)
+    | x == y = 1 + compareNextNumber
+    | otherwise = compareNextNumber
+    where compareNextNumber = countElement x ys
```
Comment (alex28sh) [inline Main.hs:39]: you should use fhigher order unctions discussed in lecture

File: Main.hs [hunk 4]
```diff
@@ -11,28 +11,47 @@ import qualified Data.List as L
...
 
-dotProduct :: [Int] -> [Int] -> Int 
-dotProduct = undefined
+sumList [] = 0
+sumList (h:t) = h + sumList t
```
Comment (alex28sh) [inline Main.hs:42]: you should use fhigher order unctions discussed in lecture instead this

File: Main.hs [hunk 5]
```diff
@@ -11,28 +11,47 @@ import qualified Data.List as L
...
+dotProduct :: [Int] -> [Int] -> Int
+dotProduct xs ys = sumList (zipWith (*) xs ys)
 
 applyAll :: [a -> b] -> a -> [b]
-applyAll = undefined
+applyAll [] x = []
+applyAll [f] x = [f x]
+applyAll (f:fs) x = f x : applyAll fs x
```
Comment (alex28sh) [inline Main.hs:49]: same

File: Main.hs [hunk 6]
```diff
@@ -11,28 +11,47 @@ import qualified Data.List as L
...
 
-interleave :: [a] -> [a] -> [a] 
-interleave = undefined
+interleave :: [a] -> [a] -> [a]
+interleave (x:xs) (y:ys) = x : y : interleave xs ys
+interleave xs [] = []
+interleave [] ys = []
```
Comment (alex28sh) [inline Main.hs:54]: you can implement this in one line using higher-order functions discussed in the lecture

**Review by alex28sh — CHANGES_REQUESTED** [review]
3/8

File: Main.hs [hunk 7]
```diff
@@ -11,28 +11,32 @@ import qualified Data.List as L
...
-powerByIndex = undefined
+powerByIndex xs = zipWith (^) xs [0..]
+
 
 productOfDifference :: [Int] -> Int
-productOfDifference = undefined
+productOfDifference xs
+    | null xs = error "The list is empty"
+    | xs == [head xs] = error "The list is too short"
+    | otherwise = foldl (*) 1 (zipWith (-) xs (tail xs))
```
Comment (kajigor) [inline Main.hs:24]: This is a very contrived way to do something trivial. Just pattern match on the list instead of using guards.

File: Main.hs [hunk 8]
```diff
@@ -11,28 +11,32 @@ import qualified Data.List as L
...
+countElement x = length . filter (x==)
 
-dotProduct :: [Int] -> [Int] -> Int 
-dotProduct = undefined
+dotProduct :: [Int] -> [Int] -> Int
+dotProduct xs ys = foldl (+) 0 (zipWith (*) xs ys)
 
 applyAll :: [a -> b] -> a -> [b]
-applyAll = undefined
+applyAll fs x = map (\f -> f x) fs
```
Comment (kajigor) [inline Main.hs:36]: You can use operator sections to avoid a lambda-function. Could you figure out how?

**Review by kajigor — CHANGES_REQUESTED** [review]

**Comment by kajigor** [comment]
ok.

**Pull request closed without merging**